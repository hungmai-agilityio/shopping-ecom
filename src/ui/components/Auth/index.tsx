import Image from 'next/image';

// Constants
import { inter, popping } from '@/constants';

interface AuthFormProps {
  title: string;
  subtitle: React.ReactNode;
  children: React.ReactNode;
}

const AuthForm = ({ title, subtitle, children }: AuthFormProps) => {
  return (
    <div className="flex my-10">
      <Image
        src="/image-form.webp"
        alt="img-form"
        width={805}
        height={781}
        className="md:block hidden"
      />
      <div className="xl:w-form w-full my-auto xl:mx-36 mx-14">
        <div className={`${popping.className}`}>
          <h1 className={`text-4xl font-medium ${inter.className}`}>{title}</h1>
          <p className="text-base mt-5">{subtitle}</p>
          {children}
        </div>
      </div>
    </div>
  );
};

export default AuthForm;
