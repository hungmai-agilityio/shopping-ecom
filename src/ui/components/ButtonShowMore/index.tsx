'use client';

import { useState, useEffect, memo, useCallback } from 'react';

// Components
import {
  Button,
  CardProductAction,
  SkeletonProductList
} from '@/ui/components';

// Interfaces
import { IProduct, IUser } from '@/interface';

// Libs
import { getProductLimit } from '@/libs';

interface ButtonShowMoreProps {
  queryKey?: string;
  user: IUser;
  startQuery: number;
}

const ButtonShowMore = memo(
  ({ queryKey = '', user, startQuery }: ButtonShowMoreProps) => {
    const [products, setProducts] = useState<IProduct[]>([]);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const [hasMoreData, setHasMoreData] = useState<boolean>(true);
    const [start, setStart] = useState<number>(startQuery);
    const limit = 4;

    useEffect(() => {
      setProducts([]);
      setHasMoreData(true);
      setStart(startQuery);
    }, [queryKey, startQuery]);

    /**
     *  Handle show more product when click button
     *  Callback when fetching limited data from the last data previously fetched
     **/
    const handleShowMore = useCallback(async () => {
      setIsLoading(true);
      const { data } = await getProductLimit(queryKey, start, limit);

      if (!data.length) {
        setHasMoreData(false);
        setIsLoading(false);
        return;
      }

      setProducts((prev) => [...prev, ...data]);
      setStart((prevStart) => prevStart + limit);
      setIsLoading(false);
    }, [start]);

    return (
      <div className="w-full mt-5">
        <div className="flex flex-wrap lg:gap-x-32 gap-14 lg:justify-normal justify-center">
          {products.map((product) => (
            <CardProductAction
              key={product.id}
              id={product.id}
              name={product.name}
              image={product.image}
              price={product.price}
              ratings={product.ratings}
              colors={product.colors || []}
              isDiscount={product.discount !== undefined}
              discount={product.discount}
              isNew={product.isNew}
              user={user}
              originalPrice={product.originalPrice}
              isNewProduct={product.isNew}
            />
          ))}
          {isLoading && <SkeletonProductList />}
        </div>

        <div className="mt-5 flex justify-center">
          <Button onClick={handleShowMore} disabled={isLoading || !hasMoreData}>
            {isLoading
              ? 'Loading...'
              : hasMoreData
                ? 'Show More'
                : 'No More Data'}
          </Button>
        </div>
      </div>
    );
  }
);

export default ButtonShowMore;
