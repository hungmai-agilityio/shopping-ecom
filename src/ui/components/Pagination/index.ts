export { default as PaginationProduct } from '@/ui/components/Pagination/Products';
