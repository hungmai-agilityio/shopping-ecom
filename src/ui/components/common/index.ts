export { default as Avatar } from '@/ui/components/common/Avatar';
export { default as Button } from '@/ui/components/common/Button';
export { default as Icon } from '@/ui/components/common/Icon';
export { default as Social } from '@/ui/components/common/Icon/Social';
export { default as Card } from '@/ui/components/common/Card';
export { default as CardPromo } from '@/ui/components/common/Card/Promo';
export { default as CardSolution } from '@/ui/components/common/Card/Solution';
export { default as CardProduct } from '@/ui/components/common/Card/Product';
export { default as Tag } from '@/ui/components/common/Tag';
export { default as Badge } from '@/ui/components/common/Badge';
export { default as Breadcrumb } from '@/ui/components/common/Breadcrumb';
export { default as Navbar } from '@/ui/components/common/Navbar';
export { default as Input } from '@/ui/components/common/Input';
export { default as SendEmail } from '@/ui/components/common/Input/SendEmail';
export { default as Logo } from '@/ui/components/common/Logo';
export { default as List } from '@/ui/components/common/List';
export { default as Countdown } from '@/ui/components/common/Countdown';
export { default as Rating } from '@/ui/components/common/Rating';
export { default as SizePicker } from '@/ui/components/common/SizePicker';
export { default as ColorPicker } from '@/ui/components/common/ColorPicker';
export { default as Carousel } from '@/ui/components/common/Carousel';
export { default as Quantity } from '@/ui/components/common/Quantity';
export { default as Dropdown } from '@/ui/components/common/Dropdown';
export { default as Table } from '@/ui/components/common/Table';

export * from '@/ui/components/common/Table/Body';
export * from '@/ui/components/common/Table/Cell';
export * from '@/ui/components/common/Table/Head';
export * from '@/ui/components/common/Table/Row';
export * from '@/ui/components/common/Typo';
