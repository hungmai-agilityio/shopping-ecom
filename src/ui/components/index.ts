export * from '@/ui/components/common';

export { default as Categories } from '@/ui/components/Categories';
