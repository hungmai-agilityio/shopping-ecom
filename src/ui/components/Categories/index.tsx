'use client';
import { clsx } from 'clsx';
import { memo } from 'react';

// Constants
import { categoryLabels, popping } from '@/constants';

// Interfaces
import { ICategory } from '@/interface';

// Components
import { Icon } from '@/ui/components';

interface CategoriesProps {
  categories: ICategory[];
  selected?: string;
  onSelect?: (category: string) => void;
}

const Categories = memo(
  ({ categories, selected, onSelect }: CategoriesProps) => {
    const handleClick = (category: string) => {
      if (onSelect) onSelect(category);
    };

    return (
      <div className={clsx('flex gap-5', popping.className)}>
        {categories.map((category) => (
          <div
            key={category.id}
            onClick={handleClick.bind(null, category.title)}
            className={clsx(
              'cursor-pointer group flex flex-col justify-center items-center w-44 h-36 border border-black rounded-sm transition duration-300',
              {
                'bg-primary text-white border-transparent':
                  selected === category.title,
                'hover:bg-primary hover:border-transparent':
                  selected !== category.title
              }
            )}
          >
            <Icon
              src={category.image}
              alt={category.title}
              width={56}
              height={56}
              className={clsx('transition duration-300', {
                'brightness-0 invert': selected === category.title,
                'group-hover:brightness-0 group-hover:invert':
                  selected !== category.title
              })}
            />
            <span
              className={clsx(
                'mt-5 text-dark transition duration-300 capitalize',
                {
                  'text-white': selected === category.title,
                  'group-hover:text-white': selected !== category.title
                }
              )}
            >
              {categoryLabels[category.title] || category.title}
            </span>
          </div>
        ))}
      </div>
    );
  }
);

export default Categories;
