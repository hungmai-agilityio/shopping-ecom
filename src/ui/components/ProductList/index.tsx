'use client';

import { useCallback, useEffect, useState } from 'react';
import { useQuery } from '@tanstack/react-query';
import { useRouter } from 'next/navigation';
import { v4 as uuidv4 } from 'uuid';

// Constants
import { END_POINT, QUERY, SIZE } from '@/constants';

// Interfaces
import { IProduct, IUser, IWishlist } from '@/interface';

// Libs
import { getProductLimit, getUserWishList } from '@/libs';

// Components
import { Button, CardProduct } from '@/ui/components';

// Hooks
import { useAddToWishlist, useRemoveFromWishlist } from '@/hooks';

interface ProductListProps {
  products: IProduct[];
  user: IUser;
  isShowMore?: boolean;
  isNewProduct?: boolean;
  isDiscount?: boolean;
  query?: string;
}

const ProductList = ({
  products,
  user,
  isNewProduct,
  isShowMore,
  isDiscount,
  query = ''
}: ProductListProps) => {
  const [currentProducts, setCurrentProducts] = useState<IProduct[]>(products);
  const [start, setStart] = useState<number>(products.length);
  const addToWishlist = useAddToWishlist();
  const removeFromWishlist = useRemoveFromWishlist();

  const router = useRouter();

  const {
    data: items,
    isLoading,
    isError
  } = useQuery({
    queryKey: [QUERY.PRODUCTS, query, start],
    queryFn: () => getProductLimit(query, start, 4),
    enabled: start > products.length
  });

  const { data: wishlist = [] } = useQuery<IWishlist[]>({
    queryKey: [QUERY.WISHLIST],
    queryFn: () => getUserWishList(user.id),
    enabled: !!user
  });

  const handleShowMore = () => {
    setStart((prev) => prev + 4);
  };

  useEffect(() => {
    if (items?.data && Array.isArray(items.data)) {
      setCurrentProducts((prev) => [...prev, ...items.data]);
    }
  }, [items]);

  if (isError) {
    return (
      <p className="text-primary font-medium text-base">
        Unable to load products! Try later
      </p>
    );
  }

  /**
   * Check if this product is in your favorites list?
   * @param {string} productId
   */
  const isProductInWishlist = (productId: string) =>
    wishlist.some((item: IWishlist) => item.productId === productId);

  /**
   * Handle toggle add/remove product with Wishlist
   * @param {IProduct} product
   */
  const handleToggleFavorite = useCallback(
    (product: IProduct) => {
      if (isProductInWishlist(product.id)) {
        const wishlistItem = wishlist.find(
          (item: IWishlist) => item.productId === product.id
        );
        if (wishlistItem) {
          removeFromWishlist.mutate(wishlistItem.id);
        }
        return;
      }

      const newItem: IWishlist = {
        id: uuidv4(),
        userId: user.id,
        productId: product.id
      };

      addToWishlist.mutate(newItem);
    },
    [removeFromWishlist, addToWishlist]
  );

  const handleRedirectPreview = (id: string) => {
    router.push(`${END_POINT.PRODUCT}/${id}`);
  };

  const isButtonDisabled = isLoading || items?.data.length === 0;

  return (
    <>
      <div className="flex flex-wrap lg:gap-x-32 gap-14 lg:justify-normal justify-center">
        {currentProducts.map((item) => (
          <CardProduct
            key={item.id}
            src={item.image}
            alt={item.name}
            icon={
              isProductInWishlist(item.id) ? '/heart-red.svg' : '/heart.svg'
            }
            id={item.id}
            name={item.name}
            price={item.price}
            rating={item.ratings}
            onAdd={() => alert(123)}
            onIconClick={handleToggleFavorite.bind(null, item)}
            onView={handleRedirectPreview}
            colors={item.colors}
            isNewProduct={isNewProduct ? item.isNew : undefined}
            discount={isDiscount ? item.discount : undefined}
            isShowAction={!!user}
          />
        ))}
      </div>
      {isLoading && <p>...Loading</p>}
      {isShowMore && (
        <div className="mt-10 text-center">
          <Button
            onClick={handleShowMore}
            size={SIZE.MEDIUM}
            disabled={isButtonDisabled}
          >
            View All Products
          </Button>
        </div>
      )}
    </>
  );
};

export default ProductList;
