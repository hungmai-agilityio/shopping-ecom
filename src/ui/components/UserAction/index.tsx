'use client';

import Link from 'next/link';

// Constants
import { END_POINT } from '@/constants';

// Components
import { Badge, Icon, UserDropdown } from '@/ui/components';

// Hooks
import { useCartData, useWishlistData } from '@/hooks';

interface UserActionProps {
  user: string;
}

const UserAction = ({ user }: UserActionProps) => {
  const { data: cart } = useCartData(user);
  const { data: wishlist } = useWishlistData(user);

  return (
    <div className="flex gap-8 items-center">
      <div className="lg:flex hidden gap-8 items-center">
        <Link href={END_POINT.WISHLIST}>
          <Badge quantity={wishlist?.length || 0}>
            <Icon
              src="/heart.svg"
              alt="heart-icon"
              width={32}
              height={32}
              priority
            />
          </Badge>
        </Link>
        <Link href={END_POINT.CART}>
          <Badge quantity={cart?.length || 0}>
            <Icon
              src="/cart.svg"
              alt="cart-icon"
              width={32}
              height={32}
              priority
            />
          </Badge>
        </Link>
      </div>
      <UserDropdown cartLength={cart.length} wishlistLength={wishlist.length} />
    </div>
  );
};

export default UserAction;
