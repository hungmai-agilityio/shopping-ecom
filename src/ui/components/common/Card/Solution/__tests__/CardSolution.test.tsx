import { render, screen } from '@testing-library/react';

// Components
import { CardSolution } from '@/ui/components';

jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => <img {...props} />
}));

describe('CardSolution component', () => {
  const defaultProps = {
    title: 'Card Title',
    desc: 'Description card',
    src: '/icon_bag.svg',
    alt: 'icon bags'
  };

  test('renders CardSolution with default props', () => {
    render(<CardSolution {...defaultProps} />);
    const card = screen.getByText('Card Title');
    expect(card).toBeInTheDocument();
  });

  test('renders CardSolution with isBorder props', () => {
    render(<CardSolution {...defaultProps} isBorder />);

    const container =
      screen.getByText('Card Title').parentElement?.parentElement;
    const bgSpan = container?.querySelector('span');
    expect(bgSpan).toHaveClass('bg-black');
    expect(bgSpan).toHaveClass('group-hover:bg-white');
  });

  test('should be render match to snapshot', () => {
    const { container } = render(<CardSolution {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
