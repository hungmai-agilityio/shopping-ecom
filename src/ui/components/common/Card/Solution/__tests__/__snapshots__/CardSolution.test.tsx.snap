// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardSolution component should be render match to snapshot 1`] = `
<div>
  <div
    class="rounded-sm overflow-hidden className"
  >
    <div
      class="flex flex-col w-card-sm h-card-sm items-center justify-center rounded-sm p-6 transition-all duration-300 text-dark truncate"
    >
      <div
        class="relative mb-4 rounded-full transition-all duration-300 shadow-card"
      >
        <span
          class="absolute inset-0 bg-black rounded-full transition-all duration-300"
        />
        <img
          alt="icon bags"
          class="relative z-10 rounded-full transition-all duration-300 p-2"
          height="40"
          src="/icon_bag.svg"
          width="40"
        />
      </div>
      <h3
        class="text-lg mt-2 font-semibold"
      >
        Card Title
      </h3>
      <p
        class="text-sm mt-2"
      >
        Description card
      </p>
    </div>
  </div>
</div>
`;
